internal/sync2/cycle.go

```diff
@@ -0,0 +1,166 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information
+
+package sync2
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"golang.org/x/sync/errgroup"
+)
+
+// Cycle implements a controllable recurring event.
+//
+// Cycle control methods don't have any effect after the cycle has completed.
+type Cycle struct {
+	interval time.Duration
+
+	ticker  *time.Ticker
+	control chan interface{}
+	quit    chan struct{}
+
+	init sync.Once
+}
+
+type (
+	// cycle control messages
+	cyclePause    struct{}
+	cycleContinue struct{}
+	cycleTrigger  struct {
+		done chan struct{}
+	}
+)
+
+// NewCycle creates a new cycle with the specified interval.
+func NewCycle(interval time.Duration) *Cycle {
+	cycle := &Cycle{}
+	cycle.SetInterval(interval)
+	return cycle
+}
+
+// SetInterval allows to change the interval before starting.
+func (cycle *Cycle) SetInterval(interval time.Duration) {
+	cycle.interval = interval
+}
+
+func (cycle *Cycle) initialize() {
+	cycle.init.Do(func() {
+		cycle.quit = make(chan struct{})
+		cycle.control = make(chan interface{})
+	})
+}
+
+// Start runs the specified function with an errgroup
+func (cycle *Cycle) Start(ctx context.Context, group *errgroup.Group, fn func(ctx context.Context) error) {
+	group.Go(func() error {
+		return cycle.Run(ctx, fn)
+	})
+}
+
+// Run runs the specified function.
+func (cycle *Cycle) Run(ctx context.Context, fn func(ctx context.Context) error) error {
+	cycle.initialize()
+	defer close(cycle.quit)
+
+	currentInterval := cycle.interval
+	cycle.ticker = time.NewTicker(currentInterval)
+	if err := fn(ctx); err != nil {
```

Comment:
There's the possibility that this function take longer to complete than interval?

Reply:
Yeah, hmm... however this is inline more how the loop in general works, since it starts to tick immediately when running the func.